src/Bedrock.Framework.Experimental/Protocols/Kafka/StrategyPayloadWriter.cs

```diff
@@ -0,0 +1,264 @@
+﻿#nullable enable
+#pragma warning disable CA1815 // Override equals and operator equals on value types
+
+using System;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.IO.Pipelines;
+using System.Linq;
+using System.Runtime.CompilerServices;
+
+namespace Bedrock.Framework.Experimental.Protocols.Kafka
+{
+    public class BigEndianStrategy : IPayloadWriterStrategy
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void WriteByte(Span<byte> destination, byte value)
+        {
+            destination[0] = value;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void WriteInt16(Span<byte> destination, short value)
+        {
+            BinaryPrimitives.WriteInt16BigEndian(destination, value);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void WriteInt32(Span<byte> destination, int value)
+        {
+            BinaryPrimitives.WriteInt32BigEndian(destination, value);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void WriteInt64(Span<byte> destination, long value)
+        {
+            BinaryPrimitives.WriteInt64BigEndian(destination, value);
+        }
+    }
+    public class LittleEndianStrategy : IPayloadWriterStrategy
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void WriteByte(Span<byte> destination, byte value)
+        {
+            destination[0] = value;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void WriteInt16(Span<byte> destination, short value)
+        {
+            BinaryPrimitives.WriteInt16LittleEndian(destination, value);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void WriteInt32(Span<byte> destination, int value)
+        {
+            BinaryPrimitives.WriteInt32LittleEndian(destination, value);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void WriteInt64(Span<byte> destination, long value)
+        {
+            BinaryPrimitives.WriteInt64LittleEndian(destination, value);
+        }
+    }
+
+    public ref struct StrategyPayloadWriter
```

Comment:
If performance is critical here, this could be `StrategyPayloadWriter<TStrategy> where TStrategy: struct, IPayloadWriterStrategy`. Then use `default(TStrategy)` to call the strategy methods.

Reply:
I went ahead and did this and saw a slight perf bump:
```
 

|                           Method | NumIterations |     Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------------- |-------------- |---------:|----------:|----------:|------:|--------:|-------:|------:|------:|----------:|
|  StrategyPayloadWriterWritesByte |           100 | 4.879 us | 0.1939 us | 0.1718 us |  1.00 |    0.00 | 0.5035 |     - |     - |   1.03 KB |
| StrategyPayloadWriterWritesShort |           100 | 4.813 us | 0.0469 us | 0.0416 us |  0.99 |    0.03 | 0.5493 |     - |     - |   1.13 KB |
|   StrategyPayloadWriterWritesInt |           100 | 4.810 us | 0.0470 us | 0.0417 us |  0.99 |    0.03 | 0.6409 |     - |     - |   1.32 KB |
|  StrategyPayloadWriterWritesLong |           100 | 4.916 us | 0.0517 us | 0.0458 us |  1.01 |    0.04 | 0.8316 |     - |     - |   1.71 KB |

|                           Method | NumIterations |     Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------------- |-------------- |---------:|----------:|----------:|------:|--------:|-------:|------:|------:|----------:|
|  StrategyPayloadWriterWritesByte |           100 | 4.663 us | 0.0881 us | 0.0781 us |  1.00 |    0.00 | 0.4883 |     - |     - |      1 KB |
| StrategyPayloadWriterWritesShort |           100 | 4.686 us | 0.0602 us | 0.0470 us |  1.01 |    0.03 | 0.5341 |     - |     - |   1.09 KB |
|   StrategyPayloadWriterWritesInt |           100 | 4.642 us | 0.0348 us | 0.0325 us |  1.00 |    0.02 | 0.6256 |     - |     - |   1.29 KB |
|  StrategyPayloadWriterWritesLong |           100 | 4.770 us | 0.0617 us | 0.0547 us |  1.02 |    0.02 | 0.8163 |     - |     - |   1.68 KB |
```

Top benchmarks are directly accessing the instance, and bottom is the default(IPayloadWriterStrategy) approach.

The problem with the latter is that the API is now a lot more complicated. The type constraints have to be on the extension methods, and implementing custom protocol behavior requires a little more ritual.

Do we think ~200 nanosecond perf boost is worth the more complicated API?